orchagent/txmonorch.cpp

```diff
@@ -0,0 +1,313 @@
+
+        string key = kfvKey(t);
+        string op = kfvOp(t);
+        vector<FieldValueTuple> fvs = kfvFieldsValues(t);
+
+        rc = -1;
+
+        SWSS_LOG_INFO("TX_ERR %s operation %s set %s del %s\n",
+                      key.c_str(),
+                      op.c_str(), SET_COMMAND, DEL_COMMAND);
+        if (key == TXMONORCH_KEY_CFG_PERIOD)
+        {
+            if (op == SET_COMMAND)
+            {
+                rc = handlePeriodUpdate(fvs);
+            }
+            else
+            {
+                SWSS_LOG_ERROR("Unknown operation type %s when set period\n", op.c_str());
+            }
+        }
+        else //key should be the alias of interface
+        {
+            if (op == SET_COMMAND)
+            {
+                //fetch the value which reprsents threshold
+                rc = handleThresholdUpdate(key, fvs, false);
+            }
+            else if (op == DEL_COMMAND)
+            {
+                //reset to default
+                rc = handleThresholdUpdate(key, fvs, true);
+            }
+            else
+            {
+                SWSS_LOG_ERROR("Unknown operation type %s when set threshold\n", op.c_str());
+            }
+        }
+
+        if (rc)
+        {
+            SWSS_LOG_ERROR("Handle configuration update failed index %s\n", key.c_str());
+        }
+
+        consumer.m_toSync.erase(it++);
+    }
+}
+
+int TxMonOrch::pollOnePortErrorStatistics(const string &port, TxErrorStatistics &stat)
+{
+    uint64_t txErrStatistics = 0,
+        txErrStatLasttime = tesStatistics(stat),
+        txErrStatThreshold = tesThreshold(stat);
+    int tx_error_state,
+        tx_error_state_lasttime = tesState(stat);
+
+    SWSS_LOG_ENTER();
+
+    {
+        vector<FieldValueTuple> fvs;
+        std::stringstream stream;
+        stream << "oid:0x" << std::hex << tesPortId(stat);
+        std::string keystr( stream.str() );
+        m_countersTable.get(keystr, fvs);
+        for (auto fv: fvs)
+        {
+            if (fvField(fv) == "SAI_PORT_STAT_IF_OUT_ERRORS") {
+                txErrStatistics = stoll(fvValue(fv));
+                break;
+            }
+        }
+
+        SWSS_LOG_INFO("TX_ERR_POLL: got port %s tx_err stati %ld, lasttime %ld threshold %ld\n",
+                      port.c_str(), txErrStatistics, txErrStatLasttime, txErrStatThreshold);
+    }
+
+    if (txErrStatistics - txErrStatLasttime > txErrStatThreshold)
+    {
+        tx_error_state = TXMONORCH_PORT_STATE_ERROR;
+    }
+    else
+    {
+        tx_error_state = TXMONORCH_PORT_STATE_OK;
+    }
+    if (tx_error_state != tx_error_state_lasttime)
+    {
+        tesState(stat) = tx_error_state;
+        //set status in STATE_DB
+        vector<FieldValueTuple> fvs;
+        if (tx_error_state < TXMONORCH_PORT_STATE_MAX)
+            fvs.emplace_back(TXMONORCH_FIELD_STATE_TX_STATE, tx_status_name[tx_error_state]);
+        else
+            fvs.emplace_back(TXMONORCH_FIELD_STATE_TX_STATE, "invalid");
```

Comment:
you have definitions to all other states. why not add a define also for invalid?

Reply:
Done.